File: cplusplus/design_pattern/ProxyProj/ProxySubject.cpp
#include "stdafx.h"
#include "ProxySubject.h"
#include "RealSubject.h"


ProxySubject::ProxySubject()
{
    realSubject = std::make_shared<RealSubject>();
}


ProxySubject::~ProxySubject()
{
}


void ProxySubject::request()
{
    preRequest();
    realSubject->request();
    postRequest();
}

void ProxySubject::preRequest()
{
    std::cout << "ProxySubject::preRequest()" << std::endl;
}


void ProxySubject::postRequest()
{
    std::cout << "ProxySubject::postRequest()" << std::endl;
}

File: Java/pattern/proxy/src/com/pattern/ProxySubject.java
package com.pattern;

public class ProxySubject implements Subject {
	private RealSubject realSubject;

	public ProxySubject() {
		realSubject = new RealSubject();
	}

	@Override
	public void request() {
		preRequest();
		realSubject.request();
		postRequest();
	}

	private void postRequest() {
		System.out.println("ProxySubject.postRequest()");
		
	}

	private void preRequest() {
		System.out.println("ProxySubject.preRequest()");
		
	}

}
